src/vmm/src/devices/virtio/net/metrics.rs

```diff
@@ -0,0 +1,435 @@
+//! doesn't consider them.
+//!
+//! # Design
+//! The main design goals of this system are:
+//! * To improve network device metrics by logging them at per device granularity.
+//! * Continue to provide aggregate net metrics to maintain backward compatibility.
+//! * Move NetDeviceMetrics out of from logger and decouple it.
+//! * Use lockless operations, preferably ones that don't require anything other than simple
+//!   reads/writes being atomic.
+//! * Rely on `serde` to provide the actual serialization for writing the metrics.
+//! * Since all metrics start at 0, we implement the `Default` trait via derive for all of them, to
+//!   avoid having to initialize everything by hand.
+//!
+//! * Devices could be created in any order i.e. the first device created could either be eth0 or
+//!   eth1 so if we use a vector for NetDeviceMetrics and call 1st device as net0, then net0 could
+//!   sometimes point to eth0 and sometimes to eth1 which doesn't help with analysing the metrics.
+//!   So, use Map instead of Vec to help understand which interface the metrics actually belongs to.
+//! * We use "net_$iface_id" for the metrics name instead of "net_$tap_name" to be consistent with
+//!   the net endpoint "/network-interfaces/{iface_id}".
+//!
+//! The system implements 1 types of metrics:
+//! * Shared Incremental Metrics (SharedIncMetrics) - dedicated for the metrics which need a counter
+//! (i.e the number of times an API request failed). These metrics are reset upon flush.
+//! We use NET_METRICS instead of adding an entry of NetDeviceMetrics
+//! in Net so that metrics are accessible to be flushed even from signal handlers.
+
+use std::collections::BTreeMap;
+use std::sync::{Arc, RwLock};
+
+use serde::ser::SerializeMap;
+use serde::{Serialize, Serializer};
+
+use crate::logger::{IncMetric, SharedIncMetric};
+
+/// map of network interface id and metrics
+/// this should be protected by a lock before accessing.
+#[derive(Debug)]
+pub struct NetMetricsPerDevice {
+    /// used to access per net device metrics
+    pub metrics: BTreeMap<String, Arc<NetDeviceMetrics>>,
+}
+
+impl NetMetricsPerDevice {
+    /// Allocate `NetDeviceMetrics` for net device having
+    /// id `iface_id`. Also, allocate only if it doesn't
+    /// exist to avoid overwriting previously allocated data.
+    /// lock is always initialized so it is safe the unwrap
+    /// the lock without a check.
+    pub fn alloc(iface_id: String) -> Arc<NetDeviceMetrics> {
+        if NET_METRICS.read().unwrap().metrics.get(&iface_id).is_none() {
+            NET_METRICS
+                .write()
+                .unwrap()
+                .metrics
+                .insert(iface_id.clone(), Arc::new(NetDeviceMetrics::default()));
+        }
+        NET_METRICS
+            .read()
+            .unwrap()
+            .metrics
+            .get(&iface_id)
+            .unwrap()
+            .clone()
```

Comment:
nit:
```suggestion
        Arc::clone(
            NET_METRICS
                .write()
                .unwrap()
                .metrics
                .entry(iface_id)
                .or_insert_with(|| Arc::new(NetDeviceMetrics::default())),
        )
```
avoids a clone.

Reply:
Since I have verified the PR with performance tests and changing this will mean I'll have to redo the tests again. In the interest of time to finalise this and since this is a nit, I would like to work on this later when I start adding per device metrics for other devices.
And thank you again for the feedback :)